### PR TITLE
Add precompiled contracts p2: proc_transfer_mnt

### PR DESCRIPTION
**Pull request by hanyunxu**
- [x]  gas cost to be discussed

- [x] unit test to be added

### CODE REVIEW

File: quarkchain/evm/specials.py [hunk 1]
```diff
@@ -237,6 +261,7 @@ def proc_ecpairing(ext, msg):
         b"0000000000000000000000000000000000000006": proc_ecadd,
         b"0000000000000000000000000000000000000007": proc_ecmul,
         b"0000000000000000000000000000000000000008": proc_ecpairing,
+        b"fffffffffffffffffffffffffffffffffffffff1": proc_transfer_mnt,
```
Comment (qizhou) [inline quarkchain/evm/specials.py]: Suggest to from '00000.....514b430000' where "514b43" is the hex of QKC

File: quarkchain/evm/specials.py [hunk 2]
```diff
@@ -226,6 +226,30 @@ def proc_ecpairing(ext, msg):
     return 1, msg.gas - gascost, [0] * 31 + [1 if result else 0]
 
 
+# 3 inputs: (address, token ID and value)
+def proc_transfer_mnt(ext, msg):
+    print("transfer_mnt proc", msg.gas)
+    gascost = 3  # to be discussed
+    if msg.gas < gascost:
+        return 0, 0, []
+    to = utils.int_to_addr(msg.data.extract32(0))
```
Comment (qizhou) [inline quarkchain/evm/specials.py]: missing data size check?

File: quarkchain/evm/specials.py [hunk 3]
```diff
@@ -226,6 +226,30 @@ def proc_ecpairing(ext, msg):
...
+    mnt = msg.data.extract32(32)
+    value = msg.data.extract32(64)
+    new_msg = vm.Message(
+        msg.to,
```
Comment (qizhou) [inline quarkchain/evm/specials.py]: msg.to => msg.sender?

File: quarkchain/evm/specials.py [hunk 4]
```diff
@@ -226,6 +226,30 @@ def proc_ecpairing(ext, msg):
...
+        to,
+        value,
+        msg.gas - gascost,
+        b"",
```
Comment (qizhou) [inline quarkchain/evm/specials.py]: need a way to supply message data.

File: quarkchain/evm/specials.py (hunk 2 above)
Comment (hanyunxu) [reply quarkchain/evm/specials.py]: Check added to make sure `data.size` is exactly 3 * 32 byte

File: quarkchain/evm/specials.py (hunk 4 above)
Comment (hanyunxu) [reply quarkchain/evm/specials.py]: Will make another PR later.

File: fixtures/GeneralStateTests/stPreCompiledContracts/transfer_mnt.json [hunk 5]
```diff
@@ -0,0 +1,85 @@
+{
```
Comment (ninjaahhh) [inline fixtures/GeneralStateTests/stPreCompiledContracts/transfer_mnt.json]: same comment as #655

File: quarkchain/evm/specials.py [hunk 6]
```diff
@@ -226,6 +226,33 @@ def proc_ecpairing(ext, msg):
     return 1, msg.gas - gascost, [0] * 31 + [1 if result else 0]
 
 
+# 3 inputs: (address, token ID and value)
+def proc_transfer_mnt(ext, msg):
+    print("transfer_mnt proc", msg.gas)
+    # Data must be exactly 96 bytes
+    if msg.data.size != 96:
+        return 0, 0, []
+    gas_cost = 3  # to be discussed
```
Comment (ninjaahhh) [inline quarkchain/evm/specials.py]: how about same as CALL or STATICCALL, 700?

**Comment by ninjaahhh** [comment]
recommend more test cases: 

1. transfer target is a contract and has a payable fallback method
1. use a non-default token ID as tx transfer token ID, and use `proc_transfer_mnt` to transfer default token ID

File: fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                "nonce" : "0x00",
+                "storage" : {
+                }
+            },
+            "0xb87213121fb89cbd8b877cb1bb3ff84dd2869cfa" : {
+               "balances" : {
+                    "0x8bb0": "0x00",
+                    "0x1234": "0x2"
+                },
+                "code" : "0x60806040526004361061003b576000357c01000000000000000000000000000000000000000000000000000000009004806390b4902714610040575b600080fd5b34801561004c57600080fd5b5061008d6004803603606081101561006357600080fd5b810190808035906020019092919080359060200190929190803590602001909291905050506100a3565b6040518082815260200191505060405180910390f35b60006100ad61011d565b848160006003811015156100bd57fe5b602002018181525050838160016003811015156100d657fe5b602002018181525050828160026003811015156100ef57fe5b602002018181525050602082606083600064514b430002600019f1151561011557600080fd5b509392505050565b60606040519081016040528060039060208202803883398082019150509050509056fea165627a7a72305820518e76d8064dc54076706d8f606977b7bd6368b267ea14f0d965dc08593965430029",
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json]: add a comment and link to the contract code (preferably using gist?)

File: fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                "balance" : "0x00",
+                "code" : "",
+                "nonce" : "0x00",
+                "storage" : {
+                }
+            }
+        },
+        "transaction" : {
+            "data" : [
+                "0x90b49027000000000000000000000000475638562837645827365482634856283465287600000000000000000000000000000000000000000000000000000000000012340000000000000000000000000000000000000000000000000000000000000002"
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json]: add a comment explain the data field

File: quarkchain/evm/specials.py (hunk 6 above)
Comment (ninjaahhh) [reply quarkchain/evm/specials.py]: cc @qizhou

File: fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json [hunk 9]
```diff
@@ -0,0 +1,81 @@
+{
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stTransferTokenId/transferMnt.json]: nit: naming, `transferMnt` -> `contractTransferMnt`

File: quarkchain/evm/specials.py [hunk 10]
```diff
@@ -224,6 +224,33 @@ def proc_current_mnt_id(ext, msg):
...
+        value,
+        msg.gas - gas_cost,
+        b"",
+        msg.depth + 1,
+        code_address=to,
+        static=msg.static,
+        transfer_token_id=mnt,
+        gas_token_id=msg.gas_token_id,
+    )
+    return messages.apply_msg(ext, new_msg)
```
Comment (ninjaahhh) [inline quarkchain/evm/specials.py]: `messages` looks confusing, recommend `from quarchain.evm.messages import apply_msg`

**Comment by ninjaahhh** [comment]
any update?

File: fixtures/QuarkChainStateTests/stTransferTokenId/contractTransferMntFallback.json [hunk 11]
```diff
@@ -0,0 +1,92 @@
+{
+    "contractTransferMnt" : {
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stTransferTokenId/contractTransferMntFallback.json]: test case naming: contractTransferMntFallback. also, recommend to put those test case files under `QuarkChainStateTests/stContractCallMnt`

File: fixtures/QuarkChainStateTests/stTransferTokenId/contractTransferMntFallback.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+            },
+            "data" : [
+                "0x90b490270000000000000000000000009dd1e8169e76a9226b07ab9f85cc20a5e1ed44dd00000000000000000000000000000000000000000000000000000000000012340000000000000000000000000000000000000000000000000000000000000009"
+            ],
+            "gasLimit" : [
+                "0x2dc6c0"
+            ],
+            "transferTokenId": [
+                "0x8bb0",
+                "0x1234"
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stTransferTokenId/contractTransferMntFallback.json]: not used

**Comment by ninjaahhh** [comment]
as discussed offline, let's add another test case that the callee contract only accepts QKC and we try to transfer 0x1234 token as well as 0x8bb0

File: fixtures/QuarkChainStateTests/stContractCallMnt/contractTransferMntQKCOnly.json [hunk 13]
```diff
@@ -0,0 +1,92 @@
+{
+    "contractTransferMnt" : {
+        "_info" : {
+            "comment" : "deployed contract: ",
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stContractCallMnt/contractTransferMntQKCOnly.json]: what contract?

File: fixtures/QuarkChainStateTests/stContractCallMnt/contractTransferMnt.json [hunk 14]
```diff
@@ -0,0 +1,92 @@
+{
+    "contractTransferMnt" : {
+        "_info" : {
+            "comment" : "deployed contract: https://gist.github.com/hanyunxu/5f0af6e57032672ff4fee7aa3b98713b",
```
Comment (ninjaahhh) [inline fixtures/QuarkChainStateTests/stContractCallMnt/contractTransferMnt.json]: this contract doesn't change any state when the tx failed or succeeded, right?

**Pull request merged**